# Exploring COVID-19 Data: An In-depth Analysis

<div>
<img src=https://thanhtra.com.vn/data/images/0/2021/12/05/congdinh/blue-covid-banner.jpg width="1500">
</div>

# Introduction

In the face of the global COVID-19 pandemic, understanding the data behind the numbers is crucial. This project dives deep into COVID-19 datasets, leveraging data analysis and SQL querying to extract meaningful insights. We explore key metrics such as total cases, deaths, vaccinations, and their percentages, shedding light on the pandemic's impact worldwide.

## Project Highlights

1. **Connect to the Database:**
   - Utilize SQLite to connect to an SQLite database for efficient data manipulation.

2. **Load the Dataset:**
   - Import COVID-19 datasets from CSV files into the SQLite database for easy access.

3. **Data Exploration:**
   - Preprocess `COVID_DEATHS` and `COVID_VACCINATIONS` datasets, correcting data types and creating processed tables.

4. **Global COVID-19 Overview:**
   - Analyze global COVID-19 statistics, including total cases, deaths, vaccination efforts, and related percentages.
   - Visualize trends in new COVID-19 cases, deaths, and vaccinations worldwide over time.

5. **COVID-19 Impact Across Continents:**
   - Explore COVID-19 metrics for continents, examining infection rates, mortality, and vaccination progress.

6. **COVID-19 Impact Across Income Levels:**
   - Analyze COVID-19 impact and vaccination responses based on income levels, highlighting disparities.

7. **COVID-19 Impact Across Countries:**
   - Delve into COVID-19 metrics for individual countries, focusing on population, total cases, deaths, vaccinations, and trends.
   - Visualize regional patterns, vaccination progress, and challenges faced by countries.

8. **COVID-19 Trends in Vietnam:**
   - Focus specifically on Vietnam's COVID-19 situation, examining total cases, deaths, vaccinations, and trends over time.
   - Visualize the progression of the pandemic in Vietnam, highlighting key insights.

## Skills Utilized

- **Joins:** Connecting data from multiple sources for comprehensive analysis.

- **Common Table Expressions (CTE's):** Simplifying complex queries and calculations.

- **Temporary Tables:** Storing interim results for efficient data manipulation.

- **Window Functions:** Performing computations over specified subsets of data.

- **Aggregate Functions:** Summarizing data to reveal trends and patterns.

- **Creating Views:** Organizing data for future reference and visualization.

- **Converting Data Types:** Ensuring data compatibility and accuracy.


This project not only showcases SQL proficiency but also aims to uncover trends, disparities, and progress in the fight against COVID-19. Join us on this data-driven journey as we unravel the story behind the numbers.

# Overview of the dataset

In this project, we focus on analyzing specific aspects of the extensive [COVID-19 dataset](https://ourworldindata.org/covid-deaths). This dataset is diligently maintained by [Our World in Data](https://ourworldindata.org), with daily updates to ensure it reflects the latest information relevant to the COVID-19 pandemic. It encompasses a diverse range of crucial metrics sourced from reputable institutions, guaranteeing accuracy and reliability.

Our analysis hones in on the following key variables:

- **Confirmed Cases:** Total and new confirmed cases, smoothed averages, and cases per million people.
- **Confirmed Deaths:** Total and new deaths attributed to COVID-19, smoothed averages, and deaths per million people.
- **Vaccinations:** Total doses administered, people vaccinated, fully vaccinated, and booster doses.

The dataset we're using contains 390,786 rows spanning from January 1, 2020, to April 18, 2024. It is divided into two CSV files for easier analysis:

- **CovidDeaths.csv**: Contains the data on confirmed cases and deaths.
- **CovidVaccinations.csv**: Contains the data on vaccinations.

For further insights, you can explore the complete dataset [here](https://github.com/owid/covid-19-data/tree/master/public/data).

# Connect to the database

We will use the SQLite database in this project to easily work with SQL on Google Colab.

In [1]:
import sqlite3

# Connect to an SQLite database; use ':memory:' for an in-memory database
conn = sqlite3.connect('covid_data.db')

In [2]:
%%capture
# Install ipython-sql
!pip install ipython-sql

In [3]:
# Load the SQL extension
%load_ext sql

# Create a SQLite database
%sql sqlite:///covid_data.db

# Load the dataset

We will import the dataset from CSV files into the SQLite database we've created.

In [4]:
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/Datasets/CovidDeaths.csv")
df.to_sql("COVID_DEATHS", conn, if_exists="append", chunksize=100, index=False, method="multi")

390786

In [6]:
df = pd.read_csv("/content/drive/MyDrive/Datasets/CovidVaccinations.csv")
df.to_sql("COVID_VACCINATIONS", conn, if_exists="append", chunksize=100, index=False, method="multi")

390786

# Data exploration

## Data preprocessing

Let's start with the `COVID_DEATHS` dataset, which contains the data on confirmed cases and deaths.

In [7]:
%%sql
SELECT continent, location, date, population, total_cases, new_cases, total_deaths, new_deaths
FROM COVID_DEATHS
ORDER BY location, date
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[('Asia', 'Afghanistan', '1/1/2021', 41128772, 51848.0, 0.0, 2158.0, 0.0),
 ('Asia', 'Afghanistan', '1/1/2022', 41128772, 157902.0, 0.0, 7352.0, 0.0),
 ('Asia', 'Afghanistan', '1/1/2023', 41128772, 207579.0, 257.0, 7849.0, 4.0),
 ('Asia', 'Afghanistan', '1/1/2024', 41128772, 230375.0, 0.0, 7973.0, 0.0),
 ('Asia', 'Afghanistan', '1/10/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/10/2021', 41128772, 53489.0, 780.0, 2277.0, 56.0),
 ('Asia', 'Afghanistan', '1/10/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/10/2023', 41128772, 207780.0, 0.0, 7851.0, 0.0),
 ('Asia', 'Afghanistan', '1/10/2024', 41128772, 230642.0, 0.0, 7973.0, 0.0),
 ('Asia', 'Afghanistan', '1/11/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/11/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/11/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/11/2023', 41128772, 207780.0, 0.0, 7851.0, 0.0),
 ('Asia', 'Afghanistan', '1/11/2024', 41128772, 230642.0, 0.0, 7973.0, 0.0),
 ('Asia', 'Afghanistan', '1/12/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/12/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/12/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/12/2023', 41128772, 207780.0, 0.0, 7851.0, 0.0),
 ('Asia', 'Afghanistan', '1/12/2024', 41128772, 230642.0, 0.0, 7973.0, 0.0),
 ('Asia', 'Afghanistan', '1/13/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/13/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/13/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/13/2023', 41128772, 207780.0, 0.0, 7851.0, 0.0),
 ('Asia', 'Afghanistan', '1/13/2024', 41128772, 230642.0, 0.0, 7973.0, 0.0),
 ('Asia', 'Afghanistan', '1/14/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/14/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/14/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/14/2023', 41128772, 207780.0, 0.0, 7851.0, 0.0),
 ('Asia', 'Afghanistan', '1/14/2024', 41128772, 230878.0, 236.0, 7975.0, 2.0),
 ('Asia', 'Afghanistan', '1/15/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/15/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/15/2022', 41128772, 158345.0, 0.0, 7369.0, 0.0),
 ('Asia', 'Afghanistan', '1/15/2023', 41128772, 207953.0, 173.0, 7855.0, 4.0),
 ('Asia', 'Afghanistan', '1/15/2024', 41128772, 230878.0, 0.0, 7975.0, 0.0),
 ('Asia', 'Afghanistan', '1/16/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/16/2021', 41128772, 53489.0, 0.0, 2277.0, 0.0),
 ('Asia', 'Afghanistan', '1/16/2022', 41128772, 158678.0, 333.0, 7378.0, 9.0),
 ('Asia', 'Afghanistan', '1/16/2023', 41128772, 207953.0, 0.0, 7855.0, 0.0),
 ('Asia', 'Afghanistan', '1/16/2024', 41128772, 230878.0, 0.0, 7975.0, 0.0),
 ('Asia', 'Afghanistan', '1/17/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/17/2021', 41128772, 53984.0, 495.0, 2339.0, 62.0),
 ('Asia', 'Afghanistan', '1/17/2022', 41128772, 158678.0, 0.0, 7378.0, 0.0),
 ('Asia', 'Afghanistan', '1/17/2023', 41128772, 207953.0, 0.0, 7855.0, 0.0),
 ('Asia', 'Afghanistan', '1/17/2024', 41128772, 230878.0, 0.0, 7975.0, 0.0),
 ('Asia', 'Afghanistan', '1/18/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/18/2021', 41128772, 53984.0, 0.0, 2339.0, 0.0),
 ('Asia', 'Afghanistan', '1/18/2022', 41128772, 158678.0, 0.0, 7378.0, 0.0),
 ('Asia', 'Afghanistan', '1/18/2023', 41128772, 207953.0, 0.0, 7855.0, 0.0),
 ('Asia', 'Afghanistan', '1/18/2024', 41128772, 230878.0, 0.0, 7975.0, 0.0),
 ('Asia', 'Afghanistan', '1/19/2020', 41128772, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '1/19/2021', 41128772, 53984.0, 0.0, 2339.0, 0.0),
 ('Asia', 'Afghanistan', '1/19/2022', 41128772, 158678.0, 0.0, 7378.0, 0.0),
 ('Asia', 'Afghanistan', '1/19/2023', 41128772, 207953.0,

We can observe that the dataset is updated weekly, with new cases and new deaths reported only on Sundays. Therefore, in this analysis, the dates of events (if mentioned) are only accurate in weeks.

Additionally, there seems to be an issue with the date column as the ordering does not function correctly in the previous query. Let's examine the table schema to identify the problem.

In [8]:
%%sql
PRAGMA table_info(COVID_DEATHS)

 * sqlite:///covid_data.db
Done.


[(0, 'iso_code', 'TEXT', 0, None, 0),
 (1, 'continent', 'TEXT', 0, None, 0),
 (2, 'location', 'TEXT', 0, None, 0),
 (3, 'date', 'TEXT', 0, None, 0),
 (4, 'population', 'INTEGER', 0, None, 0),
 (5, 'total_cases', 'REAL', 0, None, 0),
 (6, 'new_cases', 'REAL', 0, None, 0),
 (7, 'new_cases_smoothed', 'REAL', 0, None, 0),
 (8, 'total_deaths', 'REAL', 0, None, 0),
 (9, 'new_deaths', 'REAL', 0, None, 0),
 (10, 'new_deaths_smoothed', 'REAL', 0, None, 0),
 (11, 'total_cases_per_million', 'REAL', 0, None, 0),
 (12, 'new_cases_per_million', 'REAL', 0, None, 0),
 (13, 'new_cases_smoothed_per_million', 'REAL', 0, None, 0),
 (14, 'total_deaths_per_million', 'REAL', 0, None, 0),
 (15, 'new_deaths_per_million', 'REAL', 0, None, 0),
 (16, 'new_deaths_smoothed_per_million', 'REAL', 0, None, 0),
 (17, 'reproduction_rate', 'REAL', 0, None, 0),
 (18, 'icu_patients', 'REAL', 0, None, 0),
 (19, 'icu_patients_per_million', 'REAL', 0, None, 0),
 (20, 'hosp_patients', 'REAL', 0, None, 0),
 (21, 'hosp_patients_per_million', 'REAL', 0, None, 0),
 (22, 'weekly_icu_admissions', 'REAL', 0, None, 0),
 (23, 'weekly_icu_admissions_per_million', 'REAL', 0, None, 0),
 (24, 'weekly_hosp_admissions', 'REAL', 0, None, 0),
 (25, 'weekly_hosp_admissions_per_million', 'REAL', 0, None, 0)]

We see that the `date` column is now stored as a string, which is not the correct datatype for a date. To fix this, we will convert this column to a `DATE` datatype.

Next, we'll save the processed `date` column along with the other columns we're interested in into a temporary table called `COVID_DEATHS_PROCESSED` for future analysis.

In [9]:
%%sql
DROP TABLE IF EXISTS COVID_DEATHS_PROCESSED;

CREATE TABLE COVID_DEATHS_PROCESSED
(
    continent TEXT,
    location TEXT,
    date DATE,
    population REAL,
    total_cases REAL,
    new_cases REAL,
    total_deaths REAL,
    new_deaths REAL
);

INSERT INTO COVID_DEATHS_PROCESSED
WITH date_split AS (
    SELECT
      continent,
      location,
      SUBSTR(date, 1, INSTR(date, '/') - 1) AS month,
      SUBSTR(SUBSTR(date, INSTR(date, '/') + 1), 1, INSTR(SUBSTR(date, INSTR(date, '/') + 1), '/') - 1) AS date,
      SUBSTR(SUBSTR(date, INSTR(date, '/') + 1), INSTR(SUBSTR(date, INSTR(date, '/') + 1), '/') + 1) AS year,
      population,
      total_cases,
      new_cases,
      total_deaths,
      new_deaths
    FROM COVID_DEATHS
),
date_normalize AS (
    SELECT
      continent,
      location,
      year,
      CASE WHEN LENGTH(month) == 1 THEN '0' || month ELSE month END AS month,
      CASE WHEN LENGTH(date) == 1 THEN '0' || date ELSE date END AS date,
      population,
      total_cases,
      new_cases,
      total_deaths,
      new_deaths
    FROM date_split
)

SELECT
    continent,
    location,
    DATE(year || '-' || month || '-' || date) AS date,
    population,
    total_cases,
    new_cases,
    total_deaths,
    new_deaths
FROM date_normalize
ORDER BY location, date

 * sqlite:///covid_data.db
Done.
Done.
390786 rows affected.


[]

Let's take a look at our new table.

In [10]:
%%sql
SELECT *
FROM COVID_DEATHS_PROCESSED
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[('Asia', 'Afghanistan', '2020-01-05', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-06', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-07', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-08', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-09', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-10', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-11', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-12', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-13', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-14', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-15', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-16', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-17', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-18', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-19', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-20', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-21', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-22', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-23', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-24', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-25', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-26', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-27', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-28', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-29', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-30', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-01-31', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-01', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-02', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-03', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-04', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-05', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-06', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-07', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-08', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-09', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-10', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-11', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-12', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-13', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-14', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-15', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-16', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-17', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-18', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-19', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-20', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-21', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-22', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-23', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-24', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-25', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-26', 41128772.0, None, 0.0, None, 0.0),
 ('Asia', 'Afghanistan', '2020-02-27', 41128772.0, None, 0.0, None, 0.0),
 ('A

In [11]:
%%sql
PRAGMA table_info(COVID_DEATHS_PROCESSED)

 * sqlite:///covid_data.db
Done.


[(0, 'continent', 'TEXT', 0, None, 0),
 (1, 'location', 'TEXT', 0, None, 0),
 (2, 'date', 'DATE', 0, None, 0),
 (3, 'population', 'REAL', 0, None, 0),
 (4, 'total_cases', 'REAL', 0, None, 0),
 (5, 'new_cases', 'REAL', 0, None, 0),
 (6, 'total_deaths', 'REAL', 0, None, 0),
 (7, 'new_deaths', 'REAL', 0, None, 0)]

Everything looks good. Now, let's explore the number of unique locations and continents in our dataset using the processed table.

In [12]:
%%sql
SELECT DISTINCT continent, location
FROM COVID_DEATHS_PROCESSED
ORDER BY 1, 2

 * sqlite:///covid_data.db
Done.


[(None, 'Africa'),
 (None, 'Asia'),
 (None, 'Europe'),
 (None, 'European Union'),
 (None, 'High income'),
 (None, 'Low income'),
 (None, 'Lower middle income'),
 (None, 'North America'),
 (None, 'Oceania'),
 (None, 'South America'),
 (None, 'Upper middle income'),
 (None, 'World'),
 ('Africa', 'Algeria'),
 ('Africa', 'Angola'),
 ('Africa', 'Benin'),
 ('Africa', 'Botswana'),
 ('Africa', 'Burkina Faso'),
 ('Africa', 'Burundi'),
 ('Africa', 'Cameroon'),
 ('Africa', 'Cape Verde'),
 ('Africa', 'Central African Republic'),
 ('Africa', 'Chad'),
 ('Africa', 'Comoros'),
 ('Africa', 'Congo'),
 ('Africa', "Cote d'Ivoire"),
 ('Africa', 'Democratic Republic of Congo'),
 ('Africa', 'Djibouti'),
 ('Africa', 'Egypt'),
 ('Africa', 'Equatorial Guinea'),
 ('Africa', 'Eritrea'),
 ('Africa', 'Eswatini'),
 ('Africa', 'Ethiopia'),
 ('Africa', 'Gabon'),
 ('Africa', 'Gambia'),
 ('Africa', 'Ghana'),
 ('Africa', 'Guinea'),
 ('Africa', 'Guinea-Bissau'),
 ('Africa', 'Kenya'),
 ('Africa', 'Lesotho'),
 ('Africa', 'Liberia'),
 ('Africa', 'Libya'),
 ('Africa', 'Madagascar'),
 ('Africa', 'Malawi'),
 ('Africa', 'Mali'),
 ('Africa', 'Mauritania'),
 ('Africa', 'Mauritius'),
 ('Africa', 'Mayotte'),
 ('Africa', 'Morocco'),
 ('Africa', 'Mozambique'),
 ('Africa', 'Namibia'),
 ('Africa', 'Niger'),
 ('Africa', 'Nigeria'),
 ('Africa', 'Reunion'),
 ('Africa', 'Rwanda'),
 ('Africa', 'Saint Helena'),
 ('Africa', 'Sao Tome and Principe'),
 ('Africa', 'Senegal'),
 ('Africa', 'Seychelles'),
 ('Africa', 'Sierra Leone'),
 ('Africa', 'Somalia'),
 ('Africa', 'South Africa'),
 ('Africa', 'South Sudan'),
 ('Africa', 'Sudan'),
 ('Africa', 'Tanzania'),
 ('Africa', 'Togo'),
 ('Africa', 'Tunisia'),
 ('Africa', 'Uganda'),
 ('Africa', 'Western Sahara'),
 ('Africa', 'Zambia'),
 ('Africa', 'Zimbabwe'),
 ('Asia', 'Afghanistan'),
 ('Asia', 'Armenia'),
 ('Asia', 'Azerbaijan'),
 ('Asia', 'Bahrain'),
 ('Asia', 'Bangladesh'),
 ('Asia', 'Bhutan'),
 ('Asia', 'Brunei'),
 ('Asia', 'Cambodia'),
 ('Asia', 'China'),
 ('Asia', 'Georgia'),
 ('Asia', 'Hong Kong'),
 ('Asia', 'India'),
 ('Asia', 'Indonesia'),
 ('Asia', 'Iran'),
 ('Asia', 'Iraq'),
 ('Asia', 'Israel'),
 ('Asia', 'Japan'),
 ('Asia', 'Jordan'),
 ('Asia', 'Kazakhstan'),
 ('Asia', 'Kuwait'),
 ('Asia', 'Kyrgyzstan'),
 ('Asia', 'Laos'),
 ('Asia', 'Lebanon'),
 ('Asia', 'Macao'),
 ('Asia', 'Malaysia'),
 ('Asia', 'Maldives'),
 ('Asia', 'Mongolia'),
 ('Asia', 'Myanmar'),
 ('Asia', 'Nepal'),
 ('Asia', 'North Korea'),
 ('Asia', 'Northern Cyprus'),
 ('Asia', 'Oman'),
 ('Asia', 'Pakistan'),
 ('Asia', 'Palestine'),
 ('Asia', 'Philippines'),
 ('Asia', 'Qatar'),
 ('Asia', 'Saudi Arabia'),
 ('Asia', 'Singapore'),
 ('Asia', 'South Korea'),
 ('Asia', 'Sri Lanka'),
 ('Asia', 'Syria'),
 ('Asia', 'Taiwan'),
 ('Asia', 'Tajikistan'),
 ('Asia', 'Thailand'),
 ('Asia', 'Timor'),
 ('Asia', 'Turkey'),
 ('Asia', 'Turkmenistan'),
 ('Asia', 'United Arab Emirates'),
 ('Asia', 'Uzbekistan'),
 ('Asia', 'Vietnam'),
 ('Asia', 'Yemen'),
 ('Europe', 'Albania'),
 ('Europe', 'Andorra'),
 ('Europe', 'Austria'),
 ('Europe', 'Belarus'),
 ('Europe', 'Belgium'),
 ('Europe', 'Bosnia and Herzegovina'),
 ('Europe', 'Bulgaria'),
 ('Europe', 'Croatia'),
 ('Europe', 'Cyprus'),
 ('Europe', 'Czechia'),
 ('Europe', 'Denmark'),
 ('Europe', 'England'),
 ('Europe', 'Estonia'),
 ('Europe', 'Faeroe Islands'),
 ('Europe', 'Finland'),
 ('Europe', 'France'),
 ('Europe', 'Germany'),
 ('Europe', 'Gibraltar'),
 ('Europe', 'Greece'),
 ('Europe', 'Guernsey'),
 ('Europe', 'Hungary'),
 ('Europe', 'Iceland'),
 ('Europe', 'Ireland'),
 ('Europe', 'Isle of Man'),
 ('Europe', 'Italy'),
 ('Europe', 'Jersey'),
 ('Europe', 'Kosovo'),
 ('Europe', 'Latvia'),
 ('Europe', 'Liechtenstein'),
 ('Europe', 'Lithuania'),
 ('Europe', 'Luxembourg'),
 ('Europe', 'Malta'),
 ('Europe', 'Moldova'),
 ('Europe', 'Monaco'),
 ('Europe', 'Montenegro'),
 ('Europe', 'Netherlands'),
 ('Europe', 'North Macedonia'),
 ('Europe', 'Northern Ireland'),
 ('Europe', 'Norway'),
 ('Europe', 'Poland'),
 ('Europe', 'Portugal'),
 ('Europe', 'Romania'),
 ('

We can observe that there are rows where the `continent` is null, and the `location` is the continent itself (such as Asia, Africa, Europe, etc.), or descriptors like High income, Low income, Lower middle income, Upper middle income, or World.

For our further analysis in this project, to obtain accurate numbers for each country, we will filter out the rows where the `continent` is null. If we need to gather the numbers for each continent, we will use the rows where the `continent` is null and the `location` is the respective continent.

Let's redirect our focus to the second table, `COVID_VACCINATIONS`.

In [13]:
%%sql
SELECT continent, location, date, total_vaccinations, new_vaccinations
FROM COVID_VACCINATIONS
ORDER BY location, date
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[('Asia', 'Afghanistan', '1/1/2021', None, None),
 ('Asia', 'Afghanistan', '1/1/2022', None, None),
 ('Asia', 'Afghanistan', '1/1/2023', None, None),
 ('Asia', 'Afghanistan', '1/1/2024', None, None),
 ('Asia', 'Afghanistan', '1/10/2020', None, None),
 ('Asia', 'Afghanistan', '1/10/2021', None, None),
 ('Asia', 'Afghanistan', '1/10/2022', None, None),
 ('Asia', 'Afghanistan', '1/10/2023', None, None),
 ('Asia', 'Afghanistan', '1/10/2024', None, None),
 ('Asia', 'Afghanistan', '1/11/2020', None, None),
 ('Asia', 'Afghanistan', '1/11/2021', None, None),
 ('Asia', 'Afghanistan', '1/11/2022', 4925111.0, None),
 ('Asia', 'Afghanistan', '1/11/2023', None, None),
 ('Asia', 'Afghanistan', '1/11/2024', None, None),
 ('Asia', 'Afghanistan', '1/12/2020', None, None),
 ('Asia', 'Afghanistan', '1/12/2021', None, None),
 ('Asia', 'Afghanistan', '1/12/2022', None, None),
 ('Asia', 'Afghanistan', '1/12/2023', None, None),
 ('Asia', 'Afghanistan', '1/12/2024', None, None),
 ('Asia', 'Afghanistan', '1/13/2020', None, None),
 ('Asia', 'Afghanistan', '1/13/2021', None, None),
 ('Asia', 'Afghanistan', '1/13/2022', 4954809.0, None),
 ('Asia', 'Afghanistan', '1/13/2023', None, None),
 ('Asia', 'Afghanistan', '1/13/2024', None, None),
 ('Asia', 'Afghanistan', '1/14/2020', None, None),
 ('Asia', 'Afghanistan', '1/14/2021', None, None),
 ('Asia', 'Afghanistan', '1/14/2022', None, None),
 ('Asia', 'Afghanistan', '1/14/2023', None, None),
 ('Asia', 'Afghanistan', '1/14/2024', None, None),
 ('Asia', 'Afghanistan', '1/15/2020', None, None),
 ('Asia', 'Afghanistan', '1/15/2021', None, None),
 ('Asia', 'Afghanistan', '1/15/2022', None, None),
 ('Asia', 'Afghanistan', '1/15/2023', None, None),
 ('Asia', 'Afghanistan', '1/15/2024', None, None),
 ('Asia', 'Afghanistan', '1/16/2020', None, None),
 ('Asia', 'Afghanistan', '1/16/2021', None, None),
 ('Asia', 'Afghanistan', '1/16/2022', None, None),
 ('Asia', 'Afghanistan', '1/16/2023', 12603486.0, None),
 ('Asia', 'Afghanistan', '1/16/2024', None, None),
 ('Asia', 'Afghanistan', '1/17/2020', None, None),
 ('Asia', 'Afghanistan', '1/17/2021', None, None),
 ('Asia', 'Afghanistan', '1/17/2022', 4985047.0, None),
 ('Asia', 'Afghanistan', '1/17/2023', None, None),
 ('Asia', 'Afghanistan', '1/17/2024', None, None),
 ('Asia', 'Afghanistan', '1/18/2020', None, None),
 ('Asia', 'Afghanistan', '1/18/2021', None, None),
 ('Asia', 'Afghanistan', '1/18/2022', None, None),
 ('Asia', 'Afghanistan', '1/18/2023', None, None),
 ('Asia', 'Afghanistan', '1/18/2024', None, None),
 ('Asia', 'Afghanistan', '1/19/2020', None, None),
 ('Asia', 'Afghanistan', '1/19/2021', None, None),
 ('Asia', 'Afghanistan', '1/19/2022', 5004050.0, None),
 ('Asia', 'Afghanistan', '1/19/2023', None, None),
 ('Asia', 'Afghanistan', '1/19/2024', None, None),
 ('Asia', 'Afghanistan', '1/2/2021', None, None),
 ('Asia', 'Afghanistan', '1/2/2022', None, None),
 ('Asia', 'Afghanistan', '1/2/2023', None, None),
 ('Asia', 'Afghanistan', '1/2/2024', None, None),
 ('Asia', 'Afghanistan', '1/20/2020', None, None),
 ('Asia', 'Afghanistan', '1/20/2021', None, None),
 ('Asia', 'Afghanistan', '1/20/2022', None, None),
 ('Asia', 'Afghanistan', '1/20/2023', None, None),
 ('Asia', 'Afghanistan', '1/20/2024', None, None),
 ('Asia', 'Afghanistan', '1/21/2020', None, None),
 ('Asia', 'Afghanistan', '1/21/2021', None, None),
 ('Asia', 'Afghanistan', '1/21/2022', None, None),
 ('Asia', 'Afghanistan', '1/21/2023', None, None),
 ('Asia', 'Afghanistan', '1/21/2024', None, None),
 ('Asia', 'Afghanistan', '1/22/2020', None, None),
 ('Asia', 'Afghanistan', '1/22/2021', None, None),
 ('Asia', 'Afghanistan', '1/22/2022', None, None),
 ('Asia', 'Afghanistan', '1/22/2023', 12672379.0, None),
 ('Asia', 'Afghanistan', '1/22/2024', None, None),
 ('Asia', 'Afghanistan', '1/23/2020', None, None),
 ('Asia', 'Afghanistan', '1/23/2021', None, None),
 ('Asia', 'Afghanistan', '1/23/2022', None, None),
 ('Asia', 'Afghanistan', '1/23/2023', None, None),
 ('Asia', 'Afghanistan', '1/23/2024', None, None)

We've identified an issue with the data column, much like the `COVID_DEATHS` dataset. To address this, we'll preprocess it as we did earlier with the `COVID_DEATHS` dataset and then save it in a temporary table named `COVID_VACCINATIONS_PROCESSED`.

In [14]:
%%sql
DROP TABLE IF EXISTS COVID_VACCINATIONS_PROCESSED;

CREATE TABLE COVID_VACCINATIONS_PROCESSED
(
    continent TEXT,
    location TEXT,
    date DATE,
    total_vaccinations REAL,
    new_vaccinations REAL
);

INSERT INTO COVID_VACCINATIONS_PROCESSED
WITH date_split AS (
    SELECT
      continent,
      location,
      SUBSTR(date, 1, INSTR(date, '/') - 1) AS month,
      SUBSTR(SUBSTR(date, INSTR(date, '/') + 1), 1, INSTR(SUBSTR(date, INSTR(date, '/') + 1), '/') - 1) AS date,
      SUBSTR(SUBSTR(date, INSTR(date, '/') + 1), INSTR(SUBSTR(date, INSTR(date, '/') + 1), '/') + 1) AS year,
      total_vaccinations,
      new_vaccinations
    FROM COVID_VACCINATIONS
),
date_normalize AS (
    SELECT
      continent,
      location,
      year,
      CASE WHEN LENGTH(month) == 1 THEN '0' || month ELSE month END AS month,
      CASE WHEN LENGTH(date) == 1 THEN '0' || date ELSE date END AS date,
      total_vaccinations,
      new_vaccinations
    FROM date_split
)

SELECT
    continent,
    location,
    DATE(year || '-' || month || '-' || date) AS date,
    total_vaccinations,
    new_vaccinations
FROM date_normalize
ORDER BY location, date

 * sqlite:///covid_data.db
Done.
Done.
390786 rows affected.


[]

Let's take a look at our new table.

In [15]:
%%sql
SELECT *
FROM COVID_VACCINATIONS_PROCESSED
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[('Asia', 'Afghanistan', '2020-01-05', None, None),
 ('Asia', 'Afghanistan', '2020-01-06', None, None),
 ('Asia', 'Afghanistan', '2020-01-07', None, None),
 ('Asia', 'Afghanistan', '2020-01-08', None, None),
 ('Asia', 'Afghanistan', '2020-01-09', None, None),
 ('Asia', 'Afghanistan', '2020-01-10', None, None),
 ('Asia', 'Afghanistan', '2020-01-11', None, None),
 ('Asia', 'Afghanistan', '2020-01-12', None, None),
 ('Asia', 'Afghanistan', '2020-01-13', None, None),
 ('Asia', 'Afghanistan', '2020-01-14', None, None),
 ('Asia', 'Afghanistan', '2020-01-15', None, None),
 ('Asia', 'Afghanistan', '2020-01-16', None, None),
 ('Asia', 'Afghanistan', '2020-01-17', None, None),
 ('Asia', 'Afghanistan', '2020-01-18', None, None),
 ('Asia', 'Afghanistan', '2020-01-19', None, None),
 ('Asia', 'Afghanistan', '2020-01-20', None, None),
 ('Asia', 'Afghanistan', '2020-01-21', None, None),
 ('Asia', 'Afghanistan', '2020-01-22', None, None),
 ('Asia', 'Afghanistan', '2020-01-23', None, None),
 ('Asia', 'Afghanistan', '2020-01-24', None, None),
 ('Asia', 'Afghanistan', '2020-01-25', None, None),
 ('Asia', 'Afghanistan', '2020-01-26', None, None),
 ('Asia', 'Afghanistan', '2020-01-27', None, None),
 ('Asia', 'Afghanistan', '2020-01-28', None, None),
 ('Asia', 'Afghanistan', '2020-01-29', None, None),
 ('Asia', 'Afghanistan', '2020-01-30', None, None),
 ('Asia', 'Afghanistan', '2020-01-31', None, None),
 ('Asia', 'Afghanistan', '2020-02-01', None, None),
 ('Asia', 'Afghanistan', '2020-02-02', None, None),
 ('Asia', 'Afghanistan', '2020-02-03', None, None),
 ('Asia', 'Afghanistan', '2020-02-04', None, None),
 ('Asia', 'Afghanistan', '2020-02-05', None, None),
 ('Asia', 'Afghanistan', '2020-02-06', None, None),
 ('Asia', 'Afghanistan', '2020-02-07', None, None),
 ('Asia', 'Afghanistan', '2020-02-08', None, None),
 ('Asia', 'Afghanistan', '2020-02-09', None, None),
 ('Asia', 'Afghanistan', '2020-02-10', None, None),
 ('Asia', 'Afghanistan', '2020-02-11', None, None),
 ('Asia', 'Afghanistan', '2020-02-12', None, None),
 ('Asia', 'Afghanistan', '2020-02-13', None, None),
 ('Asia', 'Afghanistan', '2020-02-14', None, None),
 ('Asia', 'Afghanistan', '2020-02-15', None, None),
 ('Asia', 'Afghanistan', '2020-02-16', None, None),
 ('Asia', 'Afghanistan', '2020-02-17', None, None),
 ('Asia', 'Afghanistan', '2020-02-18', None, None),
 ('Asia', 'Afghanistan', '2020-02-19', None, None),
 ('Asia', 'Afghanistan', '2020-02-20', None, None),
 ('Asia', 'Afghanistan', '2020-02-21', None, None),
 ('Asia', 'Afghanistan', '2020-02-22', None, None),
 ('Asia', 'Afghanistan', '2020-02-23', None, None),
 ('Asia', 'Afghanistan', '2020-02-24', None, None),
 ('Asia', 'Afghanistan', '2020-02-25', None, None),
 ('Asia', 'Afghanistan', '2020-02-26', None, None),
 ('Asia', 'Afghanistan', '2020-02-27', None, None),
 ('Asia', 'Afghanistan', '2020-02-28', None, None),
 ('Asia', 'Afghanistan', '2020-02-29', None, None),
 ('Asia', 'Afghanistan', '2020-03-01', None, None),
 ('Asia', 'Afghanistan', '2020-03-02', None, None),
 ('Asia', 'Afghanistan', '2020-03-03', None, None),
 ('Asia', 'Afghanistan', '2020-03-04', None, None),
 ('Asia', 'Afghanistan', '2020-03-05', None, None),
 ('Asia', 'Afghanistan', '2020-03-06', None, None),
 ('Asia', 'Afghanistan', '2020-03-07', None, None),
 ('Asia', 'Afghanistan', '2020-03-08', None, None),
 ('Asia', 'Afghanistan', '2020-03-09', None, None),
 ('Asia', 'Afghanistan', '2020-03-10', None, None),
 ('Asia', 'Afghanistan', '2020-03-11', None, None),
 ('Asia', 'Afghanistan', '2020-03-12', None, None),
 ('Asia', 'Afghanistan', '2020-03-13', None, None),
 ('Asia', 'Afghanistan', '2020-03-14', None, None),
 ('Asia', 'Afghanistan', '2020-03-15', None, None),
 ('Asia', 'Afghanistan', '2020-03-16', None, None),
 ('Asia', 'Afghanistan', '2020-03-17', None, None),
 ('Asia', 'Afghanistan', '2020-03-18', None, None),
 ('Asia', 'Afghanistan', '2020-03-19', None, None),
 ('Asia', 'Afghanistan', '2020-03-20', None, None),
 ('Asia', 'Afghanistan', '2020-03-21', None, Non

We've identified another issue with our table: the `total_vaccinations` column has data in only a few rows, while most entries are empty. This could lead to a disjointed line chart if we aim to display total vaccinations on our dashboard. Additionally, the `new_vaccinations` column has many null values, which is unexpected given the increase in `total_vaccinations`.

To address these issues, we'll take the following steps:
- For every null value in the `total_vaccinations` column, we will replace it with the nearest non-null value of `total_vaccinations` for that country.
- We will then calculate the `new_vaccinations` by subtracting two consecutive `total_vaccinations` values.

Finally, we will save the processed data in a new temporary table called `COVID_VACCINATIONS_PROCESSED_V2`.

In [16]:
%%sql
DROP TABLE IF EXISTS COVID_VACCINATIONS_PROCESSED_V2;

CREATE TABLE COVID_VACCINATIONS_PROCESSED_V2
(
    continent TEXT,
    location TEXT,
    date DATE,
    total_vaccinations REAL,
    new_vaccinations REAL
);

INSERT INTO COVID_VACCINATIONS_PROCESSED_V2
WITH total_vaccinations_portion AS (
    SELECT
        continent,
        location,
        date,
        total_vaccinations,
        SUM(CASE WHEN total_vaccinations IS NULL THEN 0 ELSE 1 END) OVER (PARTITION BY continent, location ORDER BY date) AS total_vaccinations_partition,
        new_vaccinations
    FROM COVID_VACCINATIONS_PROCESSED
),
new_total_vaccinations AS (
    SELECT
        continent,
        location,
        date,
        FIRST_VALUE(total_vaccinations) OVER(PARTITION BY total_vaccinations_partition, continent, location) AS total_vaccinations,
        new_vaccinations
    FROM total_vaccinations_portion
)

SELECT
    continent,
    location,
    date,
    total_vaccinations,
    total_vaccinations - LAG(total_vaccinations) OVER(PARTITION BY continent, location ORDER BY date) AS new_vaccinations
FROM new_total_vaccinations
ORDER BY location, date

 * sqlite:///covid_data.db
Done.
Done.
390786 rows affected.


[]

Let's once more review our processed table.

In [17]:
%%sql
SELECT *
FROM COVID_VACCINATIONS_PROCESSED_V2
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[('Asia', 'Afghanistan', '2020-01-05', None, None),
 ('Asia', 'Afghanistan', '2020-01-06', None, None),
 ('Asia', 'Afghanistan', '2020-01-07', None, None),
 ('Asia', 'Afghanistan', '2020-01-08', None, None),
 ('Asia', 'Afghanistan', '2020-01-09', None, None),
 ('Asia', 'Afghanistan', '2020-01-10', None, None),
 ('Asia', 'Afghanistan', '2020-01-11', None, None),
 ('Asia', 'Afghanistan', '2020-01-12', None, None),
 ('Asia', 'Afghanistan', '2020-01-13', None, None),
 ('Asia', 'Afghanistan', '2020-01-14', None, None),
 ('Asia', 'Afghanistan', '2020-01-15', None, None),
 ('Asia', 'Afghanistan', '2020-01-16', None, None),
 ('Asia', 'Afghanistan', '2020-01-17', None, None),
 ('Asia', 'Afghanistan', '2020-01-18', None, None),
 ('Asia', 'Afghanistan', '2020-01-19', None, None),
 ('Asia', 'Afghanistan', '2020-01-20', None, None),
 ('Asia', 'Afghanistan', '2020-01-21', None, None),
 ('Asia', 'Afghanistan', '2020-01-22', None, None),
 ('Asia', 'Afghanistan', '2020-01-23', None, None),
 ('Asia', 'Afghanistan', '2020-01-24', None, None),
 ('Asia', 'Afghanistan', '2020-01-25', None, None),
 ('Asia', 'Afghanistan', '2020-01-26', None, None),
 ('Asia', 'Afghanistan', '2020-01-27', None, None),
 ('Asia', 'Afghanistan', '2020-01-28', None, None),
 ('Asia', 'Afghanistan', '2020-01-29', None, None),
 ('Asia', 'Afghanistan', '2020-01-30', None, None),
 ('Asia', 'Afghanistan', '2020-01-31', None, None),
 ('Asia', 'Afghanistan', '2020-02-01', None, None),
 ('Asia', 'Afghanistan', '2020-02-02', None, None),
 ('Asia', 'Afghanistan', '2020-02-03', None, None),
 ('Asia', 'Afghanistan', '2020-02-04', None, None),
 ('Asia', 'Afghanistan', '2020-02-05', None, None),
 ('Asia', 'Afghanistan', '2020-02-06', None, None),
 ('Asia', 'Afghanistan', '2020-02-07', None, None),
 ('Asia', 'Afghanistan', '2020-02-08', None, None),
 ('Asia', 'Afghanistan', '2020-02-09', None, None),
 ('Asia', 'Afghanistan', '2020-02-10', None, None),
 ('Asia', 'Afghanistan', '2020-02-11', None, None),
 ('Asia', 'Afghanistan', '2020-02-12', None, None),
 ('Asia', 'Afghanistan', '2020-02-13', None, None),
 ('Asia', 'Afghanistan', '2020-02-14', None, None),
 ('Asia', 'Afghanistan', '2020-02-15', None, None),
 ('Asia', 'Afghanistan', '2020-02-16', None, None),
 ('Asia', 'Afghanistan', '2020-02-17', None, None),
 ('Asia', 'Afghanistan', '2020-02-18', None, None),
 ('Asia', 'Afghanistan', '2020-02-19', None, None),
 ('Asia', 'Afghanistan', '2020-02-20', None, None),
 ('Asia', 'Afghanistan', '2020-02-21', None, None),
 ('Asia', 'Afghanistan', '2020-02-22', None, None),
 ('Asia', 'Afghanistan', '2020-02-23', None, None),
 ('Asia', 'Afghanistan', '2020-02-24', None, None),
 ('Asia', 'Afghanistan', '2020-02-25', None, None),
 ('Asia', 'Afghanistan', '2020-02-26', None, None),
 ('Asia', 'Afghanistan', '2020-02-27', None, None),
 ('Asia', 'Afghanistan', '2020-02-28', None, None),
 ('Asia', 'Afghanistan', '2020-02-29', None, None),
 ('Asia', 'Afghanistan', '2020-03-01', None, None),
 ('Asia', 'Afghanistan', '2020-03-02', None, None),
 ('Asia', 'Afghanistan', '2020-03-03', None, None),
 ('Asia', 'Afghanistan', '2020-03-04', None, None),
 ('Asia', 'Afghanistan', '2020-03-05', None, None),
 ('Asia', 'Afghanistan', '2020-03-06', None, None),
 ('Asia', 'Afghanistan', '2020-03-07', None, None),
 ('Asia', 'Afghanistan', '2020-03-08', None, None),
 ('Asia', 'Afghanistan', '2020-03-09', None, None),
 ('Asia', 'Afghanistan', '2020-03-10', None, None),
 ('Asia', 'Afghanistan', '2020-03-11', None, None),
 ('Asia', 'Afghanistan', '2020-03-12', None, None),
 ('Asia', 'Afghanistan', '2020-03-13', None, None),
 ('Asia', 'Afghanistan', '2020-03-14', None, None),
 ('Asia', 'Afghanistan', '2020-03-15', None, None),
 ('Asia', 'Afghanistan', '2020-03-16', None, None),
 ('Asia', 'Afghanistan', '2020-03-17', None, None),
 ('Asia', 'Afghanistan', '2020-03-18', None, None),
 ('Asia', 'Afghanistan', '2020-03-19', None, None),
 ('Asia', 'Afghanistan', '2020-03-20', None, None),
 ('Asia', 'Afghanistan', '2020-03-21', None, Non

Finally, let's create a view that combines the processed `COVID_DEATHS` and processed `COVID_VACCINATIONS` tables to facilitate easier analysis of the data in both tables later. Let's name it `COVID_COMBINE_VIEW`.

In [18]:
%%sql
CREATE VIEW COVID_COMBINE_VIEW
AS
SELECT
    dae.*,
    vac.total_vaccinations,
    vac.new_vaccinations
FROM COVID_DEATHS_PROCESSED dae
JOIN COVID_VACCINATIONS_PROCESSED_V2 vac
    ON IFNULL(dae.continent, "") = IFNULL(vac.continent, "")
    AND dae.location = vac.location
    AND dae.date = vac.date
ORDER BY 1, 2, 3

 * sqlite:///covid_data.db
Done.


[]

So now we can query from the view.

In [19]:
%%sql
SELECT *
FROM COVID_COMBINE_VIEW
LIMIT 1000 -- due to limited resources

 * sqlite:///covid_data.db
Done.


[(None, 'Africa', '2020-01-05', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-06', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-07', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-08', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-09', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-10', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-11', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-12', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-13', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-14', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-15', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-16', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-17', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-18', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-19', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-20', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-21', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-22', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-23', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-24', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-25', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-26', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-27', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-28', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-29', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-30', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-01-31', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-01', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-02', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-03', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-04', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-05', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-06', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-07', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-08', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-09', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-10', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-11', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-12', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-13', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-14', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-15', 1426736614.0, None, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-16', 1426736614.0, 1.0, 1.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-17', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-18', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-19', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-20', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-21', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-22', 1426736614.0, 1.0, 0.0, None, 0.0, None, None),
 (None, 'Africa', '2020-02-23', 142673

Ensure that the created view has the same number of rows as the original tables.

In [20]:
%%sql
SELECT 'COVID_COMBINE_VIEW' AS TABLE_NAME, COUNT(*) AS ROW_COUNT
FROM COVID_COMBINE_VIEW

UNION ALL
SELECT 'COVID_DEATHS' AS TABLE_NAME, COUNT(*) AS ROW_COUNT
FROM COVID_DEATHS

UNION ALL
SELECT 'COVID_VACCINATIONS' AS TABLE_NAME, COUNT(*) AS ROW_COUNT
FROM COVID_VACCINATIONS

 * sqlite:///covid_data.db
Done.


[('COVID_COMBINE_VIEW', 390786),
 ('COVID_DEATHS', 390786),
 ('COVID_VACCINATIONS', 390786)]

Everything looks good. Let's delve into our analysis.

## Global COVID-19 Overview

Let's explore global COVID-19 statistics.

In [21]:
# Let's look at the global numbers
%%sql
SELECT
  MAX(population) AS global_population,
  MAX(total_cases) AS global_total_cases,
  MAX(total_deaths) AS global_total_deaths,
  MAX(total_cases) / MAX(population) AS global_covid_percentage,
  MAX(total_deaths) / MAX(total_cases) AS global_death_percentage,
  MAX(total_vaccinations) AS global_total_vaccinations,
  MAX(total_vaccinations) / MAX(population) AS global_avg_vaccination
FROM COVID_COMBINE_VIEW
WHERE location == 'World'

 * sqlite:///covid_data.db
Done.


[(7975105024.0, 775251765.0, 7043660.0, 0.0972089725046861, 0.00908564200431069, 13570830469.0, 1.7016491228843282)]

**Global COVID-19 Overview**

1. **COVID-19 Impact:**
   - The global COVID percentage stands at 9.72%, indicating a significant portion of the global population affected by the virus.
   - The global death percentage is 0.91%, reflecting the impact of COVID-19 on mortality worldwide.

2. **Vaccination Efforts:**
   - There have been a total of 13.57 billion vaccinations administered globally.
   - The average vaccination rate is 1.70 doses per person, indicating progress in vaccination campaigns.

3. **Global Comparison:**
   - These global figures provide a snapshot of the scale of the COVID-19 pandemic and the response through vaccinations.
   - While progress has been made with vaccinations, the COVID percentage highlights the ongoing challenges in controlling the spread.

This analysis provides an overview of the global COVID-19 situation, including total cases, deaths, vaccination efforts, and related percentages. It offers insights into the scale of the pandemic and the progress in vaccination campaigns worldwide.

Let's explore the trend of new COVID-19 cases, deaths, and vaccinations worldwide over time.

In [22]:
%%sql
SELECT
  location,
  date,
  new_cases AS global_new_cases,
  new_deaths AS global_new_deaths,
  new_vaccinations AS global_new_vaccinations
FROM COVID_COMBINE_VIEW
WHERE location == 'World'
  AND NOT (new_cases == 0 AND new_deaths == 0)
ORDER BY date

 * sqlite:///covid_data.db
Done.


[('World', '2020-01-05', 2.0, 3.0, None),
 ('World', '2020-01-12', 45.0, 1.0, None),
 ('World', '2020-01-19', 90.0, 2.0, None),
 ('World', '2020-01-26', 1896.0, 56.0, None),
 ('World', '2020-02-02', 12538.0, 310.0, None),
 ('World', '2020-02-09', 23059.0, 545.0, None),
 ('World', '2020-02-16', 31734.0, 864.0, None),
 ('World', '2020-02-23', 9578.0, 692.0, None),
 ('World', '2020-03-01', 8272.0, 519.0, None),
 ('World', '2020-03-08', 20207.0, 650.0, None),
 ('World', '2020-03-15', 60446.0, 2654.0, None),
 ('World', '2020-03-22', 171320.0, 9698.0, None),
 ('World', '2020-03-29', 351479.0, 22285.0, None),
 ('World', '2020-04-05', 466289.0, 36847.0, None),
 ('World', '2020-04-12', 524359.0, 47675.0, None),
 ('World', '2020-04-19', 507006.0, 49959.0, None),
 ('World', '2020-04-26', 540268.0, 44409.0, None),
 ('World', '2020-05-03', 545638.0, 41130.0, None),
 ('World', '2020-05-10', 581533.0, 38923.0, None),
 ('World', '2020-05-17', 610590.0, 36165.0, None),
 ('World', '2020-05-24', 696716.0, 34748.0, None),
 ('World', '2020-05-31', 729003.0, 33368.0, None),
 ('World', '2020-06-07', 844663.0, 34882.0, None),
 ('World', '2020-06-14', 902650.0, 35276.0, None),
 ('World', '2020-06-21', 1031412.0, 37449.0, None),
 ('World', '2020-06-28', 1158117.0, 36729.0, None),
 ('World', '2020-07-05', 1328647.0, 36368.0, None),
 ('World', '2020-07-12', 1429050.0, 38795.0, None),
 ('World', '2020-07-19', 1574155.0, 42559.0, None),
 ('World', '2020-07-26', 1742871.0, 44784.0, None),
 ('World', '2020-08-02', 1839236.0, 44825.0, None),
 ('World', '2020-08-09', 1838664.0, 45689.0, None),
 ('World', '2020-08-16', 1885276.0, 44688.0, None),
 ('World', '2020-08-23', 1823522.0, 42711.0, None),
 ('World', '2020-08-30', 1875228.0, 41466.0, None),
 ('World', '2020-09-06', 1985046.0, 40000.0, None),
 ('World', '2020-09-13', 1957986.0, 43242.0, None),
 ('World', '2020-09-20', 2096023.0, 39042.0, None),
 ('World', '2020-09-27', 2104030.0, 37718.0, None),
 ('World', '2020-10-04', 2133124.0, 37571.0, None),
 ('World', '2020-10-11', 2365164.0, 38126.0, None),
 ('World', '2020-10-18', 2598276.0, 38087.0, None),
 ('World', '2020-10-25', 3090898.0, 43015.0, None),
 ('World', '2020-11-01', 3526641.0, 49499.0, None),
 ('World', '2020-11-08', 3857588.0, 59051.0, None),
 ('World', '2020-11-15', 4049247.0, 65422.0, None),
 ('World', '2020-11-22', 4110813.0, 73535.0, None),
 ('World', '2020-11-29', 4022147.0, 76177.0, None),
 ('World', '2020-12-06', 4200839.0, 79445.0, 0.0),
 ('World', '2020-12-13', 4394481.0, 82462.0, 45620.0),
 ('World', '2020-12-20', 4542432.0, 85690.0, 115469.0),
 ('World', '2020-12-27', 4118122.0, 81791.0, 254642.0),
 ('World', '2021-01-03', 4246641.0, 86299.0, 400599.0),
 ('World', '2021-01-10', 5007618.0, 92351.0, 3235591.0),
 ('World', '2021-01-17', 4853319.0, 100970.0, 1203589.0),
 ('World', '2021-01-24', 4227124.0, 103719.0, 1529660.0),
 ('World', '2021-01-31', 3735931.0, 100661.0, 3086586.0),
 ('World', '2021-02-07', 3165479.0, 89292.0, 1654923.0),
 ('World', '2021-02-14', 2720814.0, 81496.0, 2131607.0),
 ('World', '2021-02-21', 2474264.0, 68436.0, 2178111.0),
 ('World', '2021-02-28', 2661476.0, 64006.0, 14672943.0),
 ('World', '2021-03-07', 2737571.0, 60483.0, 3763540.0),
 ('World', '2021-03-14', 2963243.0, 60687.0, 16279204.0),
 ('World', '2021-03-21', 3328321.0, 63183.0, 4439063.0),
 ('World', '2021-03-28', 3832510.0, 69086.0, 8761808.0),
 ('World', '2021-04-04', 4113875.0, 75435.0, 9969327.0),
 ('World', '2021-04-11', 4559074.0, 80379.0, 11268805.0),
 ('World', '2021-04-18', 5232835.0, 85910.0, 10214627.0),
 ('World', '2021-04-25', 5695990.0, 91431.0, 16971255.0),
 ('World', '2021-05-02', 5694041.0, 96003.0, 11031469.0),
 ('World', '2021-05-09', 5464575.0, 92233.0, 13742364.0),
 ('World', '2021-05-16', 4812039.0, 87600.0, 22248846.0),
 ('World', '2021-05-23', 4216685.0, 86189.0, 22374682.0),
 ('World', '2021-05-30', 3547504.0, 78604.0, 27335340.0),
 ('World', '2021-06-06', 3006019.0, 68450.0, 26029738.0),
 ('World', '2021-06-13', 2647895.0, 711

**Global COVID-19 Trends**

1. **COVID-19 Cases and Deaths:**
   - The data reveals fluctuating trends in new COVID-19 cases and deaths over time.
   - Peak periods are noticeable, such as in March and April 2020, as well as during the winter months of 2020 and 2021.
   - A decline in cases and deaths is observed in mid-2021, followed by a rise again in late 2021 and early 2022. Subsequent decreases and sporadic increases are seen up to March 2024.

2. **Vaccination Efforts:**
   - The timeline shows the introduction and scaling up of global vaccination efforts, particularly from December 2020 onwards.
   - Notable milestones in new vaccinations occurred in late 2020 to early 2021, with significant increases in doses administered.
   - Vaccination efforts continued throughout 2021 and 2022, with varying rates of new vaccinations per week.

3. **Recent Trends (2023 - 2024):**
   - In recent years (2023 - 2024), a decrease in new cases and deaths is observed, potentially indicating the impact of vaccination campaigns.
   - The data shows a more consistent rate of new vaccinations compared to earlier periods.

This analysis provides a detailed overview of the global COVID-19 situation, highlighting trends in new cases, deaths, and vaccination efforts from January 2020 to March 2024. The insights showcase the impact of various factors on the pandemic's trajectory and the progress made in vaccination campaigns globally.

## COVID-19 Impact Across Continents

Let's analyze COVID-19 statistics for individual continents.

In [23]:
%%sql
SELECT
  location,
  MAX(population) AS population,
  MAX(total_cases) AS total_cases,
  MAX(total_deaths) AS total_deaths,
  MAX(total_vaccinations) AS total_vaccinations,
  MAX(total_cases) / MAX(population) AS covid_percentage,
  MAX(total_deaths) / MAX(total_cases) AS death_percentage,
  MAX(total_vaccinations) / MAX(population) AS avg_vaccination
FROM COVID_COMBINE_VIEW
WHERE continent IS NULL
  AND location NOT IN ('World', 'High income', 'Upper middle income', 'Lower middle income', 'Low income')
GROUP BY location
ORDER BY location

 * sqlite:///covid_data.db
Done.


[('Africa', 1426736614.0, 13140491.0, 259095.0, 863138096.0, 0.009210172971701699, 0.01971729975691167, 0.6049736773629895),
 ('Asia', 4721383370.0, 301392304.0, 1636933.0, 9102086165.0, 0.06383559232132424, 0.005431236890507994, 1.927843060327465),
 ('Europe', 744807803.0, 252450811.0, 2099805.0, 1395261693.0, 0.3389475915573887, 0.008317679755839644, 1.8733177705443562),
 ('European Union', 450146793.0, 185619587.0, 1260979.0, 951109336.0, 0.4123534586638719, 0.006793350962471433, 2.1128870643759092),
 ('North America', 600323657.0, 124541232.0, 1661466.0, 1158127259.0, 0.2074568119177086, 0.01334069025429265, 1.929171448594104),
 ('Oceania', 45038860.0, 14894138.0, 32334.0, 87655293.0, 0.3306952707062301, 0.0021709212040334257, 1.9462147354528956),
 ('South America', 436816679.0, 68826532.0, 1354009.0, 964561963.0, 0.1575638827655663, 0.01967277677160895, 2.2081619346774075)]

**COVID-19 Impact Across Continents**

1. **Variation in COVID Impact:**
   - **Europe** and the **European Union** have the highest COVID percentages at 33.89% and 41.24%, respectively.
   - **Africa** and **Asia** have lower COVID percentages at 0.92% and 6.38%, respectively.
   - **Oceania** and **North America** show moderate COVID percentages at 33.07% and 20.75%, respectively.

2. **Differences in Death Percentages:**
   - **Europe** has a death percentage of 0.83%, slightly higher than the **European Union** at 0.68%.
   - **Asia** has the lowest death percentage among continents at 0.54%.
   - **Africa** and **South America** have higher death percentages at 1.97%.

3. **Vaccination Progress:**
   - **European Union** and **South America** have the highest average vaccination rates per person, with 2.11 and 2.21 doses respectively.
   - **Asia** and **North America** also have high average vaccination rates of 1.93 doses per person.
   - **Africa** has the lowest average vaccination rate at 0.60 doses per person.

This analysis provides a snapshot of the diverse impact of COVID-19 across continents, highlighting differences in infection rates, mortality, and vaccination efforts. It shows the importance of global collaboration and targeted responses to combat the pandemic effectively.

How about examining the COVID-19 impact across different income groups?

In [24]:
%%sql
SELECT
  location,
  MAX(population) AS population,
  MAX(total_cases) AS total_cases,
  MAX(total_deaths) AS total_deaths,
  MAX(total_vaccinations) AS total_vaccinations,
  MAX(total_cases) / MAX(population) AS covid_percentage,
  MAX(total_deaths) / MAX(total_cases) AS death_percentage,
  MAX(total_vaccinations) / MAX(population) AS avg_vaccination
FROM COVID_COMBINE_VIEW
WHERE continent IS NULL
  AND location IN ('High income', 'Upper middle income', 'Lower middle income', 'Low income')
GROUP BY location
ORDER BY location

 * sqlite:///covid_data.db
Done.


[('High income', 1250514600.0, 428541299.0, 2983098.0, 2839741192.0, 0.34269195977400024, 0.006961051378154337, 2.27085808674285),
 ('Low income', 737604900.0, 2328082.0, 48045.0, 333087284.0, 0.0031562724163030913, 0.020637159687674232, 0.4515795434656142),
 ('Lower middle income', 3432097300.0, 97535147.0, 1341389.0, 4948362891.0, 0.02841852618805417, 0.01375287823167991, 1.4417898032785958),
 ('Upper middle income', 2525921300.0, 245634014.0, 2667162.0, 5449551989.0, 0.09724531560029206, 0.010858276329759444, 2.157451219481779)]

**COVID-19 Impact Across Income Levels**

1. **COVID-19 Impact:**
   - **High Income** countries have the highest COVID percentage at 34.27%, indicating a significant portion of the population affected.
   - **Low Income** countries show the lowest COVID percentage at 0.32%, suggesting relatively lower spread.

2. **Death Rates:**
   - **Low Income** countries have the highest death percentage at 2.06%, despite lower overall cases.
   - **Upper Middle Income** countries have a relatively lower death percentage at 1.09%.

3. **Vaccination Efforts:**
   - **High Income** and **Upper Middle Income** countries have the highest average vaccination rates, with 2.27 and 2.16 doses per person, respectively.
   - **Low Income** countries have the lowest average vaccination rate at 0.45 doses per person.

4. **Economic Disparities and COVID Impact:**
   - There is a clear disparity in COVID impact and vaccination rates based on income levels.
   - Lower income countries have lower COVID percentages but higher death percentages, indicating potential challenges in healthcare infrastructure.

This analysis sheds light on the disparities in COVID-19 impact and vaccination responses based on the income levels of countries, highlighting the importance of equitable access to vaccines and healthcare resources.

## COVID-19 Impact Across Countries

Let's delve into the latest COVID-19 metrics for each country, focusing on population, total cases, total deaths, total vaccinations, and derived metrics.

In [25]:
%%sql
SELECT
  location,
  MAX(population) AS population,
  MAX(total_cases) AS total_cases,
  MAX(total_deaths) AS total_deaths,
  MAX(total_vaccinations) AS total_vaccinations,
  MAX(total_cases) / MAX(population) AS covid_percentage,
  MAX(total_deaths) / MAX(total_cases) AS death_percentage,
  MAX(total_vaccinations) / MAX(population) AS avg_vaccination
FROM COVID_COMBINE_VIEW
WHERE continent IS NOT NULL
GROUP BY location
ORDER BY location

 * sqlite:///covid_data.db
Done.


[('Afghanistan', 41128772.0, 232948.0, 7985.0, 22606931.0, 0.005663869565568357, 0.03427803629994677, 0.5496621926859377),
 ('Albania', 2842318.0, 334863.0, 3605.0, 3088966.0, 0.11781334811938707, 0.010765596676849936, 1.0867770601319064),
 ('Algeria', 44903228.0, 272017.0, 6881.0, 15267442.0, 0.006057849560392407, 0.025296213104328037, 0.3400076716088206),
 ('American Samoa', 44295.0, 8359.0, 34.0, None, 0.18871204424878654, 0.004067472185668142, None),
 ('Andorra', 79843.0, 48015.0, 159.0, 157072.0, 0.601367684080007, 0.00331146516713527, 1.9672607492203449),
 ('Angola', 35588996.0, 107357.0, 1937.0, 27722924.0, 0.003016578495218016, 0.018042605512449118, 0.7789746021494959),
 ('Anguilla', 15877.0, 3904.0, 12.0, 24604.0, 0.24589028153933362, 0.0030737704918032786, 1.5496630345783209),
 ('Antigua and Barbuda', 93772.0, 9106.0, 146.0, 136512.0, 0.09710787868446871, 0.01603338458159455, 1.4557863754638911),
 ('Argentina', 45510324.0, 10130118.0, 130845.0, 116978521.0, 0.22258945025308982, 0.01291643394479709, 2.5703732849715593),
 ('Armenia', 2780472.0, 451831.0, 8777.0, 2256919.0, 0.1625015465000187, 0.019425404631377662, 0.8117035524903685),
 ('Aruba', 106459.0, 44224.0, 292.0, 174914.0, 0.4154087489080303, 0.0066027496382054995, 1.6430174996947182),
 ('Australia', 26177410.0, 11822999.0, 25020.0, 69306345.0, 0.4516489217229665, 0.00211621433783425, 2.6475631088025895),
 ('Austria', 8939617.0, 6081287.0, 22534.0, 20468731.0, 0.6802625884308019, 0.0037054656358103144, 2.2896653178765938),
 ('Azerbaijan', 10358078.0, 835272.0, 10353.0, 13857111.0, 0.08063967079606854, 0.012394764819124788, 1.337807168472761),
 ('Bahamas', 409989.0, 38084.0, 844.0, 366359.0, 0.09289029705675031, 0.022161537653607816, 0.8935825107502884),
 ('Bahrain', 1472237.0, 696614.0, 1536.0, 3476633.0, 0.47316702405930566, 0.0022049513790994724, 2.361462862297307),
 ('Bangladesh', 171186368.0, 2049519.0, 29493.0, 361710676.0, 0.011972442805726213, 0.014390205701923231, 2.1129642519198724),
 ('Barbados', 281646.0, 110580.0, 593.0, 381743.0, 0.39262052363605376, 0.005362633387592693, 1.355400041186454),
 ('Belarus', 9534956.0, 994037.0, 7118.0, 22511000.0, 0.10425187069557532, 0.007160699249625517, 2.360891859385612),
 ('Belgium', 11655923.0, 4861011.0, 34339.0, 31469387.0, 0.4170421338576104, 0.007064168338643957, 2.699862293187764),
 ('Belize', 405285.0, 71409.0, 688.0, 510717.0, 0.17619452977534328, 0.009634639891330225, 1.2601428624301418),
 ('Benin', 13352864.0, 28036.0, 163.0, 4232541.0, 0.0020996244700762323, 0.005813953488372093, 0.3169762681623957),
 ('Bermuda', 64207.0, 18860.0, 165.0, 132042.0, 0.2937374429579329, 0.008748674443266172, 2.056504742473562),
 ('Bhutan', 782457.0, 62697.0, 21.0, 2011426.0, 0.08012836488139284, 0.00033494425570601463, 2.5706537228243853),
 ('Bolivia', 12224114.0, 1212131.0, 22387.0, 14690530.0, 0.09915900653413409, 0.018469125861808666, 1.2017664429503847),
 ('Bonaire Sint Eustatius and Saba', 27052.0, 11922.0, 41.0, 35845.0, 0.4407067869288777, 0.003439020298607616, 1.3250406624279165),
 ('Bosnia and Herzegovina', 3233530.0, 403627.0, 16388.0, 1924950.0, 0.12482550030462064, 0.04060184279049717, 0.5953091512990447),
 ('Botswana', 2630300.0, 330643.0, 2801.0, 3171667.0, 0.12570543284036043, 0.008471372447019898, 1.2058194882712998),
 ('Brazil', 215313504.0, 37519960.0, 702116.0, 486436436.0, 0.17425734709142998, 0.018713132956431722, 2.259200779157818),
 ('British Virgin Islands', 31332.0, 7392.0, 64.0, 41453.0, 0.2359249329758713, 0.008658008658008658, 1.3230243840163411),
 ('Brunei', 449002.0, 343770.0, 178.0, 1293100.0, 0.7656313334907194, 0.0005177880559676528, 2.879942628317914),
 ('Bulgaria', 6781955.0, 1329171.0, 38700.0, 4726416.0, 0.19598640804900652, 0.029115892537529033, 0.696910551603483),
 ('Burkina Faso', 22673764.0, 22120.0, 400.0, 7893609.0, 0.0009755768826031708, 0.018083182640144666, 0.3481384475907926),
 ('Burundi', 12889583.0, 54569.0, 15.0, 41538.0, 0.004233573731593955, 0.00027488134288698713, 0.003

**COVID-19 Impact Across Countries**

1. **Population vs. Total Cases:**
   - The country with the highest total cases is India, with 45,034,360 cases, likely due to its massive population of 1.42 billion.
   - Iceland has a relatively small population but a high case count compared to its size.
   - China, with its large population, has a relatively low total case count compared to countries with similar populations, indicating effective control measures.

2. **COVID-19 Impact on Mortality:**
   - Mexico has a high death count (335,011) relative to its total cases (7,709,747), indicating a significant impact on mortality.
   - Lesotho, despite a low total case count (36,138), has a relatively high death count (709), suggesting a higher death rate.

3. **Vaccination Progress:**
   - Gibraltar leads in vaccination with 628.88% of its population fully vaccinated, followed by Malta (227.66%) and Iceland (562.92%).
   - Some countries like American Samoa and Eritrea do not have vaccination data available.
   - Larger countries like India and China have made significant progress in vaccination numbers, with 2.2 billion and 3.5 billion doses administered respectively.

4. **Country Comparisons:**
   - The United States has the highest total vaccination count (4.87 billion doses), reflecting its large population.
   - European countries like Spain, Italy, and Germany have vaccination rates above 2 doses per person, showing progress in vaccination campaigns.
   - African countries like Madagascar and Mali have lower vaccination rates, likely due to challenges in distribution and access.

5. **Regional Patterns:**
   - Countries in Europe, particularly Luxembourg, Iceland, and Malta, show high vaccination rates per capita.
   - South American countries like Brazil and Argentina have relatively high total case counts and death counts.
   - Southeast Asian countries like Indonesia and Malaysia have moderate vaccination rates but relatively low case counts compared to their populations.

6. **Challenges:**
   - Some countries, like Cook Islands and Nauru, have low total case counts, likely due to their isolated locations or limited testing.
   - Several countries, including Eritrea and American Samoa, have no available data for total vaccinations, possibly due to reporting issues or lack of infrastructure.

7. **Conclusion:**
   - Vaccination rates vary significantly across countries, influenced by factors such as population size, access to vaccines, and healthcare infrastructure.
   - Higher population countries face challenges in controlling the spread of the virus, leading to higher total cases.
   - Mortality rates can vary widely, influenced by factors like healthcare capacity, age demographics, and public health measures.
   - Continuous monitoring and international cooperation are crucial to managing and mitigating the impact of COVID-19 on a global scale.

These insights provide a glimpse into the diverse impact of COVID-19 across different countries and regions, highlighting both successes and challenges in the ongoing battle against the pandemic.

## COVID-19 Trends in Vietnam

After analyzing the global, continental, and individual country numbers, let's shift our focus to the cases and deaths specifically in Vietnam. Here, we delve into key metrics including total cases, deaths, vaccinations, and percentages related to the COVID-19 situation in Vietnam.

In [26]:
%%sql
SELECT
  location,
  date,
  population,
  total_cases,
  total_deaths,
  total_vaccinations,
  total_cases / population AS covid_percentage,
  total_deaths / total_cases AS death_percentage,
  total_vaccinations / population AS avg_vaccination
FROM COVID_COMBINE_VIEW
WHERE continent IS NOT NULL
  AND location == 'Vietnam'
  AND NOT (new_cases == 0 AND new_deaths == 0)
ORDER BY location, date

 * sqlite:///covid_data.db
Done.


[('Vietnam', '2020-01-26', 98186856.0, 2.0, None, None, 2.0369325197661894e-08, None, None),
 ('Vietnam', '2020-02-02', 98186856.0, 6.0, None, None, 6.110797559298569e-08, None, None),
 ('Vietnam', '2020-02-09', 98186856.0, 13.0, None, None, 1.324006137848023e-07, None, None),
 ('Vietnam', '2020-02-16', 98186856.0, 16.0, None, None, 1.6295460158129515e-07, None, None),
 ('Vietnam', '2020-03-08', 98186856.0, 20.0, None, None, 2.0369325197661894e-07, None, None),
 ('Vietnam', '2020-03-15', 98186856.0, 53.0, None, None, 5.397871177380402e-07, None, None),
 ('Vietnam', '2020-03-22', 98186856.0, 94.0, None, None, 9.57358284290109e-07, None, None),
 ('Vietnam', '2020-03-29', 98186856.0, 174.0, None, None, 1.7721312921965849e-06, None, None),
 ('Vietnam', '2020-04-05', 98186856.0, 240.0, None, None, 2.4443190237194276e-06, None, None),
 ('Vietnam', '2020-04-12', 98186856.0, 258.0, None, None, 2.6276429504983842e-06, None, None),
 ('Vietnam', '2020-04-19', 98186856.0, 268.0, None, None, 2.729489576486694e-06, None, None),
 ('Vietnam', '2020-04-26', 98186856.0, 270.0, None, None, 2.7498589016843556e-06, None, None),
 ('Vietnam', '2020-05-10', 98186856.0, 288.0, None, None, 2.9331828284633127e-06, None, None),
 ('Vietnam', '2020-05-17', 98186856.0, 318.0, None, None, 3.2387227064282412e-06, None, None),
 ('Vietnam', '2020-05-24', 98186856.0, 325.0, None, None, 3.310015344620058e-06, None, None),
 ('Vietnam', '2020-05-31', 98186856.0, 328.0, None, None, 3.340569332416551e-06, None, None),
 ('Vietnam', '2020-06-07', 98186856.0, 329.0, None, None, 3.3507539950153815e-06, None, None),
 ('Vietnam', '2020-06-14', 98186856.0, 334.0, None, None, 3.4016773080095366e-06, None, None),
 ('Vietnam', '2020-06-21', 98186856.0, 349.0, None, None, 3.554447246992001e-06, None, None),
 ('Vietnam', '2020-06-28', 98186856.0, 355.0, None, None, 3.6155552225849865e-06, None, None),
 ('Vietnam', '2020-07-12', 98186856.0, 370.0, None, None, 3.7683251615674503e-06, None, None),
 ('Vietnam', '2020-07-19', 98186856.0, 382.0, None, None, 3.890541112753422e-06, None, None),
 ('Vietnam', '2020-07-26', 98186856.0, 418.0, None, None, 4.257188966311336e-06, None, None),
 ('Vietnam', '2020-08-02', 98186856.0, 590.0, 3.0, None, 6.008950933310259e-06, 0.005084745762711864, None),
 ('Vietnam', '2020-08-09', 98186856.0, 812.0, 10.0, None, 8.269946030250729e-06, 0.012315270935960592, None),
 ('Vietnam', '2020-08-16', 98186856.0, 951.0, 23.0, None, 9.68561413148823e-06, 0.024185068349106203, None),
 ('Vietnam', '2020-08-23', 98186856.0, 1014.0, 26.0, None, 1.032724787521458e-05, 0.02564102564102564, None),
 ('Vietnam', '2020-08-30', 98186856.0, 1040.0, 32.0, None, 1.0592049102784186e-05, 0.03076923076923077, None),
 ('Vietnam', '2020-09-06', 98186856.0, 1049.0, 35.0, None, 1.0683711066173664e-05, 0.03336510962821735, None),
 ('Vietnam', '2020-09-13', 98186856.0, 1060.0, 35.0, None, 1.0795742354760804e-05, 0.0330188679245283, None),
 ('Vietnam', '2020-09-20', 98186856.0, 1068.0, 35.0, None, 1.0877219655551451e-05, 0.03277153558052434, None),
 ('Vietnam', '2020-09-27', 98186856.0, 1069.0, 35.0, None, 1.0887404318150282e-05, 0.03274087932647334, None),
 ('Vietnam', '2020-10-04', 98186856.0, 1096.0, 35.0, None, 1.1162390208318718e-05, 0.03193430656934307, None),
 ('Vietnam', '2020-10-11', 98186856.0, 1107.0, 35.0, None, 1.1274421496905859e-05, 0.031616982836495035, None),
 ('Vietnam', '2020-10-18', 98186856.0, 1126.0, 35.0, None, 1.1467930086283646e-05, 0.03108348134991119, None),
 ('Vietnam', '2020-10-25', 98186856.0, 1160.0, 35.0, None, 1.18142086146439e-05, 0.03017241379310345, None),
 ('Vietnam', '2020-11-01', 98186856.0, 1180.0, 35.0, None, 1.2017901866620518e-05, 0.029661016949152543, None),
 ('Vietnam', '2020-11-08', 98186856.0, 1213.0, 35.0, None, 1.2353995732381939e-05, 0.028854080791426217, None),
 ('Vietnam', '2020-11-15', 98186856.0, 1265.0, 35.0, None, 1.2883598187521149e-05, 0.02766798418972332, None),
 ('Vietnam', '2020-11-22', 98186856.0, 1306.0, 35.0, None, 1.3301

**Data Analysis: COVID-19 Trends in Vietnam**

**Overview:**
- **Location**: Vietnam
- **Population**: 98,186,856

**Key Metrics:**
- **Total Cases**: The number of confirmed COVID-19 cases.
- **Total Deaths**: The number of deaths due to COVID-19.
- **Total Vaccinations**: The total number of COVID-19 vaccinations administered.
- **COVID-19 Percentage**: Percentage of the population infected with COVID-19.
- **Death Percentage**: Percentage of COVID-19 cases resulting in death.
- **Average Vaccination**: Average number of vaccinations per day.

**Insights:**

**Total Cases and Deaths:**
- The number of COVID-19 cases in Vietnam started from 2 on 2020-01-26 and gradually increased over time.
- Total cases reached 1,142,641 by 2022-09-11, with a significant increase seen in 2022.
- Total deaths started from 0 and reached 43,130 by 2022-09-11. The death percentage remained relatively low.

**Vaccination Efforts:**
- Vaccination data starts from 2021-03-07, and the total number of vaccinations reached 258,723,446 by 2022-09-11.
- Average daily vaccinations increased over time, indicating an acceleration in vaccination efforts.
- The COVID-19 percentage began to decrease noticeably after the vaccination campaign started, suggesting its effectiveness.

**Trends Over Time:**
- **COVID-19 Cases**: The number of cases showed fluctuations but generally increased over time.
- **Vaccinations**: Vaccination numbers rose steadily from the start, with a notable increase in the number of daily vaccinations.
- **COVID-19 Percentage**: After peaking in mid-2021, the COVID-19 percentage declined, likely due to vaccination efforts.

**Conclusion:**
- Vietnam's COVID-19 situation saw a significant increase in cases in 2022, likely prompting an escalation in vaccination efforts.
- The increasing vaccination rate correlates with a decrease in COVID-19 percentage, indicating the effectiveness of vaccination campaigns.
- Despite the rising cases, the death percentage remained relatively low, possibly due to the vaccination campaign's impact.

This analysis provides a snapshot of Vietnam's COVID-19 situation, highlighting the trends in cases, deaths, vaccinations, and their impact on the population.

# Conclusion

In conclusion, this project has provided a comprehensive analysis of COVID-19 data using SQL querying techniques. Here are the key takeaways:

- **Global Overview:** We explored the global impact of COVID-19, highlighting total cases, deaths, vaccination efforts, and related percentages. The data revealed significant progress in vaccination campaigns, impacting COVID-19 percentages and trends over time.

- **Continental and Income-Level Analysis:** Our analysis across continents and income levels showcased disparities in COVID-19 impact and vaccination rates. Higher income countries generally exhibited higher vaccination rates, while lower income countries faced challenges in controlling the spread.

- **Country-Specific Insights:** Delving into individual countries' data provided a nuanced view of the pandemic. We observed varied trends in total cases, deaths, and vaccination rates, reflecting regional differences in healthcare infrastructure and response strategies.

- **Vietnam's COVID-19 Trends:** Focusing on Vietnam, we analyzed its total cases, deaths, vaccinations, and trends over time. The data revealed a notable increase in cases in 2022, corresponding with an acceleration in vaccination efforts. The decreasing death percentage suggested positive impacts from these efforts.

## Key Learnings

- **Data Analysis Skills:** Utilizing SQL and data visualization tools, we gained insights into complex datasets.
- **Regional Disparities:** Highlighted the importance of equitable access to vaccines and healthcare resources across different regions and income levels.
- **Impact of Vaccination:** Demonstrated the effectiveness of vaccination campaigns in reducing COVID-19 percentages and mortality rates.



## Future Considerations


- **Continued Monitoring:** Ongoing monitoring of COVID-19 data is crucial for adapting strategies and interventions.
- **Equitable Distribution:** Ensuring equitable distribution of vaccines to all regions and income levels remains a critical challenge.
- **Healthcare Infrastructure:** Investing in healthcare infrastructure and response capabilities is essential for future pandemic preparedness.

This project underscores the power of data analysis in understanding and combating global health crises. By examining COVID-19 data through multiple lenses, we gain valuable insights that can inform policy decisions, healthcare strategies, and international collaborations in the ongoing fight against the pandemic.